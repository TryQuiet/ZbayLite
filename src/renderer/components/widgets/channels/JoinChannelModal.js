import React from 'react'
import PropTypes from 'prop-types'
import { DateTime } from 'luxon'
import * as R from 'ramda'
import { Formik, Form } from 'formik'
import { Typography } from '@material-ui/core'
import Grid from '@material-ui/core/Grid'
import { withStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'

import Modal from '../../ui/Modal/Modal'
import { AutocompleteField } from '../../ui/Autocomplete/Autocomplete'
import { errorNotification } from '../../../store/handlers/utils'
import LoadindButton from '../../ui/LoadingButton/LoadingButton'

const styles = theme => ({
  root: {
    padding: 32,
    height: '100%',
    width: '100%'
  },
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  input: {
    marginBottom: 0,
    marginTop: 0,
    width: '100%'
  },
  button: {
    width: 151,
    height: 60,
    fontSize: 16,
    backgroundColor: theme.palette.colors.zbayBlue,
    color: theme.palette.colors.white,
    '&:hover': {
      backgroundColor: theme.palette.colors.zbayBlue
    }
  },
  title: {
    marginBottom: 24
  },
  info: {
    marginTop: 8,
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px'
  },
  channelTitle: {
    lineHeight: '26px',
    fontSize: 16
  },
  channelInfo: {
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px'
  },
  option: {},
  materialOption: {
    height: 68
  },
  timeInfo: {
    marginTop: -16,
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px',
    marginBottom: 16
  },
  description: {
    fontSize: 14,
    lineHeight: '24px'
  },
  informationBox: {
    backgroundColor: theme.palette.colors.gray03,
    borderRadius: 4,
    height: 42,
    color: theme.palette.colors.trueBlack,
    letterSpacing: '0.4px',
    lineHeight: '18px',
    fontSize: 12,
    paddingLeft: 16,
    paddingRight: 16,
    marginBottom: 24,
    marginTop: 16
  }
})
export const JoinChannelModal = ({
  classes,
  open,
  handleClose,
  joinChannel,
  Channel,
  publicChannels,
  showNotification,
  users
}) => {
  const channelsArray = Array.from(Object.values(publicChannels))
  const [step, setStep] = React.useState(0)
  const [loading, setLoading] = React.useState(false)
  return (
    <Modal
      open={open}
      handleClose={() => {
        handleClose()
        setStep(0)
      }}
      title=''
      fullPage>
      <Grid className={classes.root}>
        <Formik
          onSubmit={async (values, { resetForm }) => {
            const ch = channelsArray.find(channel => channel.name === values.channel.name)
            if (ch) {
              if (!ch) {
                showNotification(errorNotification({ message: 'Channel does not exist' }))
                return
              }
              setLoading(true)
              await joinChannel(ch)
              setLoading(false)
              setStep(0)
              handleClose()
              resetForm()
              return
            }
            showNotification(
              errorNotification({
                message: 'There was an error. Please check channel URL'
              })
            )
          }}>
          {({ values, setFieldValue }) => {
            return (
              <Form className={classes.fullContainer}>
                <Grid
                  container
                  justify='flex-start'
                  direction='column'
                  className={classes.fullContainer}>
                  <Typography variant='h3' className={classes.title}>
                    {step === 0 ? 'Search for Channels' : `#${values.channel.name}`}
                  </Typography>
                  {step !== 0 && (
                    <Typography variant='caption' className={classes.timeInfo}>
                      {`Created by @${
                        users[values.channel.owner]
                          ? users[values.channel.owner].nickname
                          : 'Unnamed'
                      } on ${DateTime.fromSeconds(parseInt(values.channel.timestamp)).toFormat(
                        'LLL d, y'
                      )} `}
                    </Typography>
                  )}
                  {step === 0 ? (
                    <AutocompleteField
                      name={'channel'}
                      classes={{ option: classes.materialOption }}
                      options={channelsArray}
                      renderOption={option => {
                        const time = DateTime.fromSeconds(parseInt(option.timestamp))
                        return (
                          <Grid container direction='column' className={classes.option}>
                            <Typography variant='body1' className={classes.channelTitle}>
                              {`#${option.name}`}
                            </Typography>
                            <Typography variant='caption' className={classes.channelInfo}>
                              {`Created by @${
                                users[option.owner] ? users[option.owner].nickname : 'Unnamed'
                              } on ${time.toFormat('LLL d, y')} `}
                            </Typography>
                          </Grid>
                        )
                      }}
                      value={values.channel}
                      onChange={(e, v) => {
                        setFieldValue('channel', v)
                        setStep(1)
                      }}
                      renderInput={params => (
                        <TextField
                          {...params}
                          className={classes.input}
                          variant='outlined'
                          placeholder={'Search'}
                          margin='normal'
                        />
                      )}
                    />
                  ) : (
                    <>
                      <Typography variant='body2' className={classes.description}>
                        {`${values.channel.description}`}
                      </Typography>
                      <Grid container alignItems='center' className={classes.informationBox}>
                        After joining, it may take some time for messages to fully load.
                      </Grid>
                    </>
                  )}

                  {step !== 0 ? (
                    <LoadindButton
                      variant='contained'
                      color='primary'
                      size='large'
                      type='submit'
                      text='Join Channel'
                      classes={{ button: classes.button }}
                      inProgress={loading}
                      disabled={loading}
                    />
                  ) : (
                    <Typography variant='caption' className={classes.info}>
                      If you have an invite link, open it in a browser
                    </Typography>
                  )}
                </Grid>
              </Form>
            )
          }}
        </Formik>
      </Grid>
    </Modal>
  )
}

JoinChannelModal.propTypes = {
  classes: PropTypes.object.isRequired,
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  joinChannel: PropTypes.func.isRequired,
  showNotification: PropTypes.func.isRequired,
  publicChannels: PropTypes.object.isRequired,
  users: PropTypes.object.isRequired
}

export default R.compose(React.memo, withStyles(styles))(JoinChannelModal)
