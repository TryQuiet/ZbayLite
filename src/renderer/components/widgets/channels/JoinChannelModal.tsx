import React from 'react'
import { DateTime } from 'luxon'
import { Formik, Form } from 'formik'
import { Typography } from '@material-ui/core'
import Grid from '@material-ui/core/Grid'
import { makeStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'

import Modal from '../../ui/Modal/Modal'
import { AutocompleteField } from '../../ui/Autocomplete/Autocomplete'
import { errorNotification, IErrorNotification } from '../../../store/handlers/utils'
import LoadindButton from '../../ui/LoadingButton/LoadingButton'
import { Channel } from '../../../store/handlers/channel'
import { User } from '@zbayapp/nectar/lib/sagas/users/users.slice'

const useStyles = makeStyles((theme) => ({
  root: {
    padding: 32,
    height: '100%',
    width: '100%'
  },
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  input: {
    marginBottom: 0,
    marginTop: 0,
    width: '100%'
  },
  button: {
    width: 151,
    height: 60,
    fontSize: 16,
    backgroundColor: theme.palette.colors.zbayBlue,
    color: theme.palette.colors.white,
    '&:hover': {
      backgroundColor: theme.palette.colors.zbayBlue
    }
  },
  title: {
    marginBottom: 24
  },
  info: {
    marginTop: 8,
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px'
  },
  channelTitle: {
    lineHeight: '26px',
    fontSize: 16
  },
  channelInfo: {
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px'
  },
  option: {},
  materialOption: {
    height: 68
  },
  timeInfo: {
    marginTop: -16,
    lineHeight: '18px',
    color: theme.palette.colors.darkGray,
    letterSpacing: '0.4px',
    marginBottom: 16
  },
  description: {
    fontSize: 14,
    lineHeight: '24px'
  },
  informationBox: {
    backgroundColor: theme.palette.colors.gray03,
    borderRadius: 4,
    height: 42,
    color: theme.palette.colors.trueBlack,
    letterSpacing: '0.4px',
    lineHeight: '18px',
    fontSize: 12,
    paddingLeft: 16,
    paddingRight: 16,
    marginBottom: 24,
    marginTop: 16
  }
}))

interface JoinChannelModalProps {
  open: boolean
  handleClose: () => void
  joinChannel: () => void
  publicChannels: Channel
  showNotification: (arg: IErrorNotification) => void
  users: User
}

export const JoinChannelModal: React.FC<JoinChannelModalProps> = ({
  open,
  handleClose,
  joinChannel,
  publicChannels,
  showNotification,
  users
}) => {
  const classes = useStyles({})
  const channelsArray = Array.from(Object.values(publicChannels))
  const [step, setStep] = React.useState(0)
  const [loading, setLoading] = React.useState(false)
  return (
    <Modal
      open={open}
      handleClose={() => {
        handleClose()
        setStep(0)
      }}
      title=''
      fullPage>
      <Grid className={classes.root}>
        <Formik
          initialValues={{
            channel: {
              name: '',
              owner: '',
              timestamp: '',
              description: ''
            }
          }}
          onSubmit={async (values, { resetForm }) => {
            const ch = channelsArray.find(channel => channel.name === values.channel.name)
            if (ch) {
              if (!ch) {
                showNotification(errorNotification({ message: 'Channel does not exist' }))
                return
              }
              setLoading(true)
              await joinChannel()
              setLoading(false)
              setStep(0)
              handleClose()
              resetForm()
              return
            }
            showNotification(
              errorNotification({
                message: 'There was an error. Please check channel URL'
              })
            )
          }}>
          {({ values, setFieldValue }) => {
            return (
              <Form className={classes.fullContainer}>
                <Grid
                  container
                  justify='flex-start'
                  direction='column'
                  className={classes.fullContainer}>
                  <Typography variant='h3' className={classes.title}>
                    {step === 0 ? 'Search for Channels' : `#${values.channel.name}`}
                  </Typography>
                  {step !== 0 && (
                    <Typography variant='caption' className={classes.timeInfo}>
                      {`Created by @${users[values.channel.owner]
                        ? users[values.channel.owner].nickname
                        : 'Unnamed'
                        } on ${DateTime.fromSeconds(
                          parseInt(values.channel.timestamp)
                        ).toFormat('LLL d, y')} `}
                    </Typography>
                  )
                  }
                  {step === 0 ? (
                    <AutocompleteField
                      name={'channel'}
                      classes={{ option: classes.materialOption }}
                      options={channelsArray}
                      renderOption={option => {
                        const time = DateTime.fromSeconds(parseInt(option.timestamp))
                        return (
                          <Grid container direction='column' className={classes.option}>
                            <Typography variant='body1' className={classes.channelTitle}>
                              {`#${option.name}`}
                            </Typography>
                            <Typography
                              variant='caption'
                              className={classes.channelInfo}
                            >
                              {`Created by @${users[option.owner]
                                ? users[option.owner].nickname
                                : 'Unnamed'
                                } on ${time.toFormat('LLL d, y')} `}
                            </Typography>
                          </Grid>
                        )
                      }}
                      value={values.channel}
                      onChange={(v: string) => {
                        setFieldValue('channel', v)
                        setStep(1)
                      }}
                      renderInput={params => (
                        <TextField
                          {...params}
                          className={classes.input}
                          variant='outlined'
                          placeholder={'Search'}
                          margin='normal'
                        />
                      )}
                    />
                  ) : (
                    <>
                      <Typography variant='body2' className={classes.description}>
                        {`${values.channel.description}`}
                      </Typography>
                      <Grid container alignItems='center' className={classes.informationBox}>
                        After joining, it may take some time for messages to fully load.
                      </Grid>
                    </>
                  )}

                  {step !== 0 ? (
                    <LoadindButton
                      variant='contained'
                      color='primary'
                      size='large'
                      type='submit'
                      text='Join Channel'
                      classes={{ button: classes.button }}
                      inProgress={loading}
                      disabled={loading}
                    />
                  ) : (
                    <Typography variant='caption' className={classes.info}>
                      If you have an invite link, open it in a browser
                    </Typography>
                  )}
                </Grid>
              </Form>
            )
          }}
        </Formik >
      </Grid >
    </Modal >
  )
}

export default JoinChannelModal
