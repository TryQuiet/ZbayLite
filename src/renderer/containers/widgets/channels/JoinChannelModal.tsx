import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { publicChannels } from '@zbayapp/nectar'
import JoinChannelModalComponent from '../../../components/widgets/channels/JoinChannelModal'
import channelHandlers from '../../../store/handlers/channel'
// import notificationsHandlers from '../../../store/handlers/notifications'
import modalsHandlers from '../../../store/handlers/modals'
import modalsSelectors from '../../../store/selectors/modals'

const useJoinChannelData = () => {
  const modalName = 'joinChannel'
  const data = {
    publicChannels: useSelector(publicChannels.selectors.publicChannels),
    users: [],
    open: useSelector(modalsSelectors.open(modalName)),
    modalName
  }
  return data
}

export const useJoinChannelActions = () => {
  const dispatch = useDispatch()
  const modalName = 'joinChannel'
  const actions = {
    joinChannel: channel => {
      dispatch(channelHandlers.epics.linkChannelRedirect(channel))
    },
    handleClose: () => {
      dispatch(modalsHandlers.closeModalHandler(modalName))
    }
    // showNotification: ()  => {dispatch(notificationsHandlers.actions.enqueueSnackbar()}
  }
  return actions
}

export const JoinChannelModal = () => {
  const { publicChannels, open, users } = useJoinChannelData()
  const { joinChannel, handleClose } = useJoinChannelActions()

  return (
    <JoinChannelModalComponent
      publicChannels={publicChannels}
      joinChannel={joinChannel}
      // showNotification={showNotification}
      open={open}
      users={users}
      handleClose={handleClose}
    />
  )
}

export default JoinChannelModal
