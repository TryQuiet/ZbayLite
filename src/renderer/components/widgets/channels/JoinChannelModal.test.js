import React from 'react'
import { shallow } from 'enzyme'

import { mockClasses } from '../../../../shared/testing/mocks'
import { JoinChannelModal } from './JoinChannelModal'

describe('JoinChannelModal', () => {
  it('renders component', () => {
    const result = shallow(
      <JoinChannelModal
        open
        handleClose={() => {}}
        joinChannel={() => {}}
        showNotification={() => {}}
        publicChannels={{}}
        users={{}}
        classes={mockClasses}
      />
    )
    expect(result).toMatchSnapshot()
  })
})
