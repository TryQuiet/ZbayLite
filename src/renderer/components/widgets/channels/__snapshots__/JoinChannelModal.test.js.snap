// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinChannelModal renders component 1`] = `
<Modal
  addBorder={false}
  alignCloseLeft={false}
  canGoBack={false}
  contentWidth={600}
  fullPage={true}
  handleClose={[Function]}
  isBold={false}
  isCloseDisabled={false}
  open={true}
  title=""
>
  <WithStyles(ForwardRef(Grid))
    className="root-mock"
  >
    <Formik
      enableReinitialize={false}
      isInitialValid={false}
      onSubmit={[Function]}
      validateOnBlur={true}
      validateOnChange={true}
    >
      <Component />
    </Formik>
  </WithStyles(ForwardRef(Grid))>
</Modal>
`;
